extensions/cargo/main.ts

```diff
@@ -0,0 +1,237 @@
+import {
+  green,
+  red,
+  yellow,
+} from "https://deno.land/std@0.150.0/fmt/colors.ts";
+import { PhylumApi, PolicyEvaluationResponseRaw } from "phylum";
+
+class FileBackup {
+  readonly fileName: string;
+  fileContent: string | null;
+
+  constructor(fileName: string) {
+    this.fileName = fileName;
+    this.fileContent = null;
+  }
+
+  async backup() {
+    try {
+      this.fileContent = await Deno.readTextFile(this.fileName);
+    } catch (_e) { /* Do nothing */ }
+  }
+
+  async restoreOrDelete() {
+    try {
+      if (this.fileContent != null) {
+        await Deno.writeTextFile(this.fileName, this.fileContent);
+      } else {
+        await Deno.remove(this.fileName);
+      }
+    } catch (_e) { /* Do nothing */ }
+  }
+}
+
+// Find project root directory.
+async function findRoot(manifest: string): Promise<string | undefined> {
+  let workingDir = Deno.cwd();
+
+  // Traverse up to 32 directories to find the root directory.
+  for (let i = 0; i < 32; i++) {
+    try {
+      // Check if manifest exists at location.
+      await Deno.stat(workingDir + "/" + manifest);
+      return workingDir;
+    } catch (_e) {
+      // Pop to parent if manifest doesn't exist.
+      workingDir += "/..";
+    }
+  }
+
+  return undefined;
+}
+
+// Ensure no arguments are passed before a subcommand.
+const firstSubcommand = Deno.args.findIndex((arg) => !arg.startsWith("-"));
+if (firstSubcommand > 0) {
+  console.error(
+    `[${
+      red("phylum")
+    }] This extension does not support arguments before the first subcommand. Please open an issue if "${
+      Deno.args[0]
+    }" is not an argument.`,
+  );
+  Deno.exit(127);
+}
+
+// Ignore all commands that shouldn't be intercepted.
+if (
+  Deno.args.length == 0 ||
+  !(
+    "add".startsWith(Deno.args[0]) ||
+    "update".startsWith(Deno.args[0])
+  )
+) {
+  const cmd = new Deno.Command("cargo", { args: Deno.args });
+  const status = await cmd.spawn().status;
+  Deno.exit(status.code);
+}
+
+// Ensure we're in a Cargo root directory.
+const root = await findRoot("Cargo.toml");
+if (!root) {
+  console.error(`[${red("phylum")}] unable to find Cargo project root.`);
+  console.error(
+    `[${
+      red("phylum")
+    }] Please change to a Cargo project directory and try again.`,
+  );
+  Deno.exit(124);
+}
+
+// Store initial package manager file state.
+const packageLockBackup = new FileBackup(root + "/Cargo.lock");
+await packageLockBackup.backup();
+const manifestBackup = new FileBackup(root + "/Cargo.toml");
+await manifestBackup.backup();
+
+// Run the command in a sandbox.
+const status = PhylumApi.runSandboxed({
+  cmd: "cargo",
+  args: Deno.args,
+  exceptions: {
+    run: ["cargo", "~/.rustup"],
+    read: ["./", "~/.cargo", "~/.rustup", "/etc/passwd"],
+    write: ["./", "~/.cargo"],
+    net: true,
+  },
+});
```

Comment:
Technically I'm not sure there's any reason for this sandbox, but since we only run this for `add` and `update` I suppose it shouldn't be a problem either.